src/utils.ts

```diff
@@ -133,33 +133,330 @@ export namespace Utils {
+   *         .toString();
+   *
+   * // The result below should be equal to [clientKP1.publicKey(), clientKP2.publicKey()]
+   * Utils.verifyChallengeTxSigners(signedChallenge, serverKP.publicKey(), Networks.TESTNET, threshold, clientKP1.publicKey(), clientKP2.publicKey());
+   */
+  export function verifyChallengeTxSigners(
+    challengeTx: string,
+    serverAccountID: string,
+    networkPassphrase: string,
+    ...accountIDs: string[]
+  ): string[] {
+    // Read the transaction which validates its structure.
+    const { tx } = readChallengeTx(
+      challengeTx,
+      serverAccountID,
+      networkPassphrase,
+    );
+
+    // Ensure the server account ID is an address and not a seed.
+    const serverKP = Keypair.fromPublicKey(serverAccountID); // can throw 'Invalid Stellar public key'
+
+    // Deduplicate the client signers and ensure the server is not included
+    // anywhere we check or output the list of signers.
+    const clientSigners: string[] = new Array();
+    const clientSignersSeen = new Map<string, boolean>();
+    for (const signer of accountIDs) {
+      // Ignore the server signer if it is in the signers list. It's
+      // important when verifying signers of a challenge transaction that we
+      // only verify and return client signers. If an account has the server
+      // as a signer the server should not play a part in the authentication
+      // of the client.
+      if (signer === serverKP.publicKey()) {
+        continue;
+      }
+
+      // Deduplicate.
+      if (clientSignersSeen.get(signer) !== undefined) {
+        continue;
+      }
+
+      // Ignore non-G... account/address signers.
+      if (!StrKey.isValidEd25519PublicKey(signer)) {
+        continue;
+      }
+
+      clientSigners.push(signer);
+      clientSignersSeen.set(signer, true);
+    }
+
+    // Don't continue if none of the signers provided are in the final list.
+    if (clientSigners.length === 0) {
       throw new InvalidSep10ChallengeError(
-        "The transaction is not signed by the client",
+        "No verifiable client signers provided, at least one G... address must be provided",
       );
     }
 
-    if (!validateTimebounds(transaction)) {
-      throw new InvalidSep10ChallengeError("The transaction has expired");
+    // Verify all the transaction's signers (server and client) in one
+    // hit. We do this in one hit here even though the server signature was
+    // checked in the ReadChallengeTx to ensure that every signature and signer
+    // are consumed only once on the transaction.
+    const allSigners: string[] = [serverKP.publicKey(), ...clientSigners];
+    const allSignersFound: string[] = verifyTxMultiSignedBy(tx, ...allSigners);
+
+    // Confirm the server is in the list of signers found and remove it.
+    let serverSignerFound: boolean = false;
+    const signersFound: string[] = new Array();
+    for (const signer of allSignersFound) {
+      if (signer === serverKP.publicKey()) {
```

Comment:
I don't think we need this loop --  we can use `allSignersFound.includes(serverKP.publicKey()` to check if the server signed was found and then if

```
if (serverSignerFound &&  allSignersFound.length  == 1) {
  throw new InvalidSep10ChallengeError(
        "Transaction not signed by client(s): '" + clientSigners.join() + "'",
      );
}
```